File: src/medium/valid-sudoku/2.js
/*
 * @lc app=leetcode id=36 lang=javascript
 *
 * [36] Valid Sudoku
 */

// @lc code=start
/**
 * @param {character[][]} board
 * @return {boolean}
 */
var isValidSudoku = function (board) {
  const recordSet = new Set();
  for (let i = 0; i < board.length; i++) {
    for (let j = 0; j < board[i].length; j++) {
      const item = board[i][j];
      if (item === '.') {
        continue;
      }
      const x = `x${j}${item}`;
      const y = `y${i}${item}`;
      const xy = `${Math.floor(i / 3)}${Math.floor(j / 3)}${item}`;
      if (recordSet.has(x)) {
        return false;
      }
      if (recordSet.has(y)) {
        return false;
      }
      if (recordSet.has(xy)) {
        return false;
      }
      recordSet.add(x);
      recordSet.add(y);
      recordSet.add(xy);
    }
  }
  return true;
};
// @lc code=end

File: src/easy/valid-palindrome/2.go
// approaching 2
// RB: 100  MB: 92.88
package main

import (
	"unicode"
)

/*
 * @lc app=leetcode id=125 lang=golang
 *
 * [125] Valid Palindrome
 */

// @lc code=start
func isPalindromeA2(s string) bool {
	l, r := 0, len(s)-1
	for l < r {
		for l < r && !unicode.IsLetter(rune(s[l])) && !unicode.IsNumber(rune(s[l])) {
			l++
		}
		for l < r && !unicode.IsLetter(rune(s[r])) && !unicode.IsNumber(rune(s[r])) {
			r--
		}
		if unicode.ToLower(rune(s[l])) != unicode.ToLower(rune(s[r])) {
			return false
		}
		l++
		r--
	}
	return true
}

// @lc code=end
